File: 8. Week 8 - Colour/solutions/r/visualization.R
# Visualize Monthly Distance with Color Scale using ggplot2

if (!require(ggplot2)) install.packages("ggplot2")
if (!require(readr)) install.packages("readr")
library(ggplot2)
library(readr)

data_path <- "../../data/distance_per_month.csv"

if (file.exists(data_path)) {
  df <- read_csv(data_path, show_col_types = FALSE)
  
  # Plot
  p <- ggplot(df, aes(x = Month, y = TotalDistance, fill = TotalDistance)) +
    geom_col() +
    
    # Sequential Color Scale
    scale_fill_gradient(low = "plum", high = "rebeccapurple", name = "Distance") +
    
    # Labels
    labs(
      title = "Monthly Distance (Colored by Intensity)",
      x = "Month",
      y = "Distance (km)"
    ) +
    
    # Theme
    theme_minimal() +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1)
    )
  
  ggsave("colored_bar_chart.png", plot = p, width = 12, height = 6)
  print("Chart saved to colored_bar_chart.png")

} else {
  print(paste("Error: File not found at", data_path))
}

File: 8. Week 8 - Colour/solutions/python/visualization.py
import pandas as pd
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as mcolors
import numpy as np
import os

def visualize():
    data_path = '../../data/distance_per_month.csv'
    
    if not os.path.exists(data_path):
        print(f"Data not found at {data_path}")
        return

    df = pd.read_csv(data_path)
    
    # Setup Colors
    # Normalize distance values to 0-1 range for colormap
    norm = mcolors.Normalize(vmin=df['TotalDistance'].min(), vmax=df['TotalDistance'].max())
    # Choose a colormap (e.g., 'viridis', 'plasma', 'Blues')
    cmap = cm.get_cmap('Purples')
    
    # Create colors for each bar
    colors = [cmap(norm(val)) for val in df['TotalDistance']]
    
    plt.figure(figsize=(12, 6))
    
    # Bar Chart with custom colors
    bars = plt.bar(df['Month'], df['TotalDistance'], color=colors)
    
    plt.title('Monthly Distance (Colored by Intensity)', fontsize=16)
    plt.xlabel('Month')
    plt.ylabel('Distance (km)')
    plt.xticks(rotation=45)
    
    # Add a Colorbar to show scale
    sm = cm.ScalarMappable(cmap=cmap, norm=norm)
    sm.set_array([])
    plt.colorbar(sm, label='Distance Intensity')
    
    plt.tight_layout()
    plt.savefig('colored_bar_chart.png')
    print("Chart saved to colored_bar_chart.png")
    plt.show()

if __name__ == "__main__":
    visualize()
